<a href="https://colab.research.google.com/github/LatiefDataVisionary/feature-engineering-college-task/blob/main/P4_FeatureEng.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Tugas Mata Kuliah Feature Engineering Pengganti Pertemuan 4: Transformasi Data**


Nama          : Lathif Ramadhan

NPM           : 5231811022

Prodi/Kelas   : Sains Data/23A

## **1. Mengupload dataset `train.csv` ke colab**

In [1]:
from google.colab import files
uploaded = files.upload()

Saving train.csv to train.csv





## **2. Load Dataset**

In [3]:
import pandas as pd

data = pd.read_csv('train.csv')
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


## **3. Identifikasi Fitur yang Akan Ditransformasi**
Kolom yang akan di transformasi:
*   Numerik: `Age`, `Fare`
*   Kategorikal: `Sex`, `Embarked`

## **4. Normalisasi Fitur Numerik**
Menggunakan ***Min-Max Scaling*** untuk menormalkan fitur `Age` dan `Fare`. Disini saya menggunakan `MinMaxScaler` dari **Scikit-learn**.

In [8]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
numeric_features = ['Age', 'Fare']

# Normalisasi proses
data[numeric_features] = scaler.fit_transform(data[numeric_features])
print(data[numeric_features].head())

        Age      Fare
0  0.271174  0.014151
1  0.472229  0.139136
2  0.321438  0.015469
3  0.434531  0.103644
4  0.434531  0.015713


## **5. One-Hot Encoding Fitur Kategorikal**

**[One-Hot encoding](https://ilmudatapy.com/one-hot-encoding-di-python/)** adalah salah satu metode encoding. Metode ini merepresentasikan data bertipe kategori sebagai vektor biner yang bernilai integer, 0 dan 1, dimana semua elemen akan bernilai 0 kecuali satu elemen yang bernilai 1, yaitu elemen yang memiliki nilai kategori tersebut.


Untuk mengubah fitur `Sex` dan `Embarked` ke dalam bentuk **one-hot encoding**, gunakan `pd.get_dummies()`.

In [12]:
data = pd.get_dummies(data, columns=['Sex', 'Embarked'])
print(data.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name       Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris  0.271174      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  0.472229      1      0   
2                             Heikkinen, Miss. Laina  0.321438      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  0.434531      1      0   
4                           Allen, Mr. William Henry  0.434531      0      0   

             Ticket      Fare Cabin  Sex_female  Sex_male  Embarked_C  \
0         A/5 21171  0.014151   NaN       False      True       False   
1          PC 17599  0.139136   C85        True     False        True   
2  STON/O2. 3101282  0.015469   NaN        True     False       False   
3         

In [ ]:
print(data[['Sex_female', 'Sex_male', 'Embarked_C', 'Embarked_Q', 'Embarked_S']].head())

   Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S
0       False      True       False       False        True
1        True     False        True       False       False
2        True     False       False       False        True
3        True     False       False       False        True
4       False      True       False       False        True


Dari hasil output diatas menunjukkan bahwa proses **one-hot encoding** sudah berhasil dilakukan pada kolom `Sex` dan `Embarked`.

**Penjelasan Output:**

**1. Kolom Sex:**

*  `Sex_female`: Berisi nilai `True` jika penumpang berjenis kelamin perempuan, dan `False` jika tidak.
*  `Sex_male`: Berisi nilai `True` jika penumpang berjenis kelamin laki-laki, dan `False` jika tidak.
Karena ini adalah **one-hot encoding**, setiap baris hanya akan memiliki satu kolom yang bernilai `True` antara `Sex_female` dan `Sex_male`.

**2. Kolom Embarked:**

*  `Embarked_C`: Berisi `True` jika penumpang naik dari pelabuhan **Cherbourg**.
*  `Embarked_Q`: Berisi `True` jika penumpang naik dari pelabuhan **Queenstown**.
*  `Embarked_S`: Berisi `True` jika penumpang naik dari pelabuhan **Southampton**.
Kolom-kolom ini juga merupakan hasil dari **one-hot encoding**, di mana hanya salah satu dari kolom tersebut yang bernilai `True` untuk setiap penumpang, sesuai dengan pelabuhan tempat mereka naik.

## **6. Periksa Missing Values**

In [14]:
print(data.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Sex_female       0
Sex_male         0
Embarked_C       0
Embarked_Q       0
Embarked_S       0
dtype: int64


Berdasarkan output diatas, terlihat bahwa kolom `Age` dan `Cabin` memiliki **missing values**, khususnya:

*  `Age` memiliki **177** **missing values**.
*  `Cabin` memiliki **687** **missing values**.

Untuk mengatasi masalah ini, kamu bisa memilih salah satu pendekatan berikut:

### **Mengisi Missing Values(Imputasi/Imputation)**


Untuk `Age`, pengisian **missing values** menggunakan **median** dari kolom tersebut.

In [15]:
data['Age'].fillna(data['Age'].median(), inplace=True)
print(data['Age'].isnull().sum())  # Hasilnya harus 0

0


<ipython-input-15-b876c5102777>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(data['Age'].median(), inplace=True)


Untuk `Cabin`, karena terlalu banyak **missing values** (**687** dari **891** total penumpang), kita bisa mempertimbangkan untuk menghapus kolom tersebut dari analisis jika dirasa tidak relevan, karena mungkin terlalu banyak data yang hilang.

In [16]:
data.drop('Cabin', axis=1, inplace=True)
print(data.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name       Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris  0.271174      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  0.472229      1      0   
2                             Heikkinen, Miss. Laina  0.321438      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  0.434531      1      0   
4                           Allen, Mr. William Henry  0.434531      0      0   

             Ticket      Fare  Sex_female  Sex_male  Embarked_C  Embarked_Q  \
0         A/5 21171  0.014151       False      True       False       False   
1          PC 17599  0.139136        True     False        True       False   
2  STON/O2. 3101282  0.015469        True     False       False  

## **7. Normalisasi ulang kolom `Age` setelah Imputasi**

In [17]:
data['Age'] = scaler.fit_transform(data[['Age']])
print(data['Age'].head())

0    0.271174
1    0.472229
2    0.321438
3    0.434531
4    0.434531
Name: Age, dtype: float64


## **8. Simpan Dataset yang Sudah Bersih**

In [19]:
from google.colab import files

data.to_csv('data_train_ditransformasi.csv', index=False)
files.download('data_train_ditransformasi.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>